### PR TITLE
Admiral apis

### PR DESCRIPTION
**Pull request by vrushalijoshi**
Admiral APIs - 
1. Get all clusters admiral is watching
2. Get all se by cluster name
3. Get all se by service identity

### CODE REVIEW

File: admiral/pkg/apis/admiral/routes/handlers.go [hunk 1]
```diff
@@ -25,4 +38,94 @@ func (opts *RouteOpts) ReturnSuccessGET(w http.ResponseWriter, r *http.Request)
...
+
+		serviceEntriesByCluster, err := clusters.GetServiceEntriesByCluster(clusterName, opts.RemoteRegistry)
+
+		if err != nil {
+			log.Printf(err.Error())
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		} else {
+			if len(serviceEntriesByCluster) == 0 {
+				log.Printf(fmt.Sprintf("No service entries configured for cluster - %s", clusterName))
+				http.Error(w, fmt.Sprintf("No service entries configured for cluster - %s", clusterName), http.StatusNotFound)
```
Comment (aattuluri) [inline admiral/pkg/apis/admiral/routes/handlers.go:77]: I think the behavior should be:
The cluster doesn't exist (is not being monitored by admiral) -> 404
The cluster has 0 service entries -> empty response object with 200

File: admiral/pkg/apis/admiral/routes/handlers.go [hunk 2]
```diff
@@ -25,4 +38,94 @@ func (opts *RouteOpts) ReturnSuccessGET(w http.ResponseWriter, r *http.Request)
...
+			} else {
+				response = serviceEntriesByCluster
+
+				out, err := json.Marshal(response)
+				if err != nil {
+					log.Printf("Failed to marshall response")
```
Comment (aattuluri) [inline admiral/pkg/apis/admiral/routes/handlers.go:83]: Same as above, provide request context.

**Review by aattuluri — CHANGES_REQUESTED** [review]
Good work!
Please see comments and add unit tests please.

File: admiral/pkg/apis/admiral/routes/handlers.go (hunk 2 above)
Comment (vrushalijoshi) [reply admiral/pkg/apis/admiral/routes/handlers.go]: Adding fix for this

File: admiral/pkg/apis/admiral/routes/handlers.go (hunk 1 above)
Comment (vrushalijoshi) [reply admiral/pkg/apis/admiral/routes/handlers.go]: Makes sense, updated the response handling

File: admiral/pkg/apis/admiral/routes/handlers.go [hunk 3]
```diff
@@ -25,4 +38,94 @@ func (opts *RouteOpts) ReturnSuccessGET(w http.ResponseWriter, r *http.Request)
...
+	var clusterList []string
+
+	// loop through secret controller's c.cs.remoteClusters to access all clusters admiral is watching
+	for clusterID := range opts.RemoteRegistry.SecretController.Cs.RemoteClusters {
+		clusterList = append(clusterList, clusterID)
+	}
+
+	out, err := json.Marshal(clusterList)
+	if err != nil {
+		log.Printf("Failed to marshall response")
```
[the remark this answers is not in the document]
Comment (vrushalijoshi) [reply admiral/pkg/apis/admiral/routes/handlers.go]: done

File: admiral/pkg/apis/admiral/routes/handlers.go [hunk 4]
```diff
@@ -1,17 +1,30 @@
...
 	KubeconfigPath string
 	RemoteRegistry *clusters.RemoteRegistry
 }
 
+//type ClusterServiceEntries struct {
+//	ServiceEntries []v1alpha3.ServiceEntry `json:"ServiceEntries,omitempty"`
+//}
+
+type IdentityServiceEntry struct {
+	IdentityCname string   `json:"IdentityCname,omitempty"`
```
Comment (aattuluri) [inline admiral/pkg/apis/admiral/routes/handlers.go:24]: Can this be changed to just Cname?, IdentityCname seems a bit misleading

**Review by aattuluri — APPROVED** [review]

File: admiral/pkg/apis/admiral/routes/handlers.go (hunk 4 above)
Comment (vrushalijoshi) [reply admiral/pkg/apis/admiral/routes/handlers.go]: Sure, added fix for this

**Comment by Mengying-Li** [comment]
@vrushalijoshi  just a quick question. I'm trying to understand why the SeClusterCache needs to be a map of map. I notice this pattern many times in other caches as well. But why do we need to put host, clusterID, clusterID this format? Why twice on the cluster ID? This part is just a question for my own understanding.

**Comment by aattuluri** [comment]
> @vrushalijoshi just a quick question. I'm trying to understand why the SeClusterCache needs to be a map of map. I notice this pattern many times in other caches as well. But why do we need to put host, clusterID, clusterID this format? Why twice on the cluster ID? This part is just a question for my own understanding.

Because go doesn't support hash set, we could have simply done host, clusterID, true kind of format as well

**Comment by Mengying-Li** [comment]
> Because go doesn't support hash set, we could have simply done host, clusterID, true kind of format as well

why do we need true here? Doesn't host, clusterId and delete when not used is good enough? @aattuluri

File: admiral/pkg/apis/admiral/routes/handlers.go [hunk 5]
```diff
@@ -25,4 +38,106 @@ func (opts *RouteOpts) ReturnSuccessGET(w http.ResponseWriter, r *http.Request)
...
+	clusterName := strings.Trim(params["clustername"], " ")
+
+	var response []v1alpha3.ServiceEntry
+
+	if clusterName != "" {
+
+		serviceEntriesByCluster, err := clusters.GetServiceEntriesByCluster(clusterName, opts.RemoteRegistry)
+
+		if err != nil {
+			log.Printf(err.Error())
```
Comment (aattuluri) [inline admiral/pkg/apis/admiral/routes/handlers.go:80]: Print some request context, would be hard to understand what this error is for

File: admiral/pkg/apis/admiral/routes/handlers.go [hunk 6]
```diff
@@ -25,4 +38,106 @@ func (opts *RouteOpts) ReturnSuccessGET(w http.ResponseWriter, r *http.Request)
...
+			if strings.Contains(err.Error(), "Admiral is not monitoring cluster") {
+				http.Error(w, err.Error(), http.StatusNotFound)
+			} else {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+		} else {
+			if len(serviceEntriesByCluster) == 0 {
+				log.Printf(fmt.Sprintf("No service entries configured for cluster - %s", clusterName))
```
Comment (aattuluri) [inline admiral/pkg/apis/admiral/routes/handlers.go:88]: Print some request context

**Review by aattuluri — CHANGES_REQUESTED** [review]

**Review by aattuluri — APPROVED** [review]

File: admiral/pkg/apis/admiral/routes/handlers.go (hunk 5 above)
Comment (Mengying-Li) [reply admiral/pkg/apis/admiral/routes/handlers.go:80]: Sorry I just saw the comment. I will address that in the other branch @aattuluri

**Pull request merged**